pyqs/worker.py

```diff
@@ -267,6 +267,164 @@ def _run_hooks(self, hook_name, context):
             hook(context)
 
 
+class SimpleProcessWorker(BaseWorker):
+
+    def __init__(self, queue_url, batchsize,
+                 connection_args=None, *args, **kwargs):
+        super(SimpleProcessWorker, self).__init__(*args, **kwargs)
+        if connection_args is None:
+            connection_args = {}
+        self.connection_args = connection_args
+        self.queue_url = queue_url
+
+        sqs_queue = get_conn(**self.connection_args).get_queue_attributes(
+            QueueUrl=queue_url, AttributeNames=['All'])['Attributes']
+        self.visibility_timeout = int(sqs_queue['VisibilityTimeout'])
+
+        self.batchsize = batchsize
+        self._messages_to_process_before_shutdown = 100
+        self.messages_processed = 0
+
+    def run(self):
+        # Set the child process to not receive any keyboard interrupts
+        signal.signal(signal.SIGINT, signal.SIG_IGN)
+
+        logger.info(
+            "Running SimpleProcessWorker: {}, pid: {}".format(
+                self.queue_url, os.getpid()))
+
+        while not self.should_exit.is_set() and self.parent_is_alive():
+            self.read_message()
+
+            if self.messages_processed \
+                    >= self._messages_to_process_before_shutdown:
+                self.shutdown()
+
+    def read_message(self):
+        messages = self._get_connection().receive_message(
+            QueueUrl=self.queue_url,
+            MaxNumberOfMessages=self.batchsize,
+            WaitTimeSeconds=LONG_POLLING_INTERVAL,
+        ).get('Messages', [])
+
+        logger.debug(
+            "Successfully got {} messages from SQS queue {}".format(
+                len(messages), self.queue_url))  # noqa
+
+        start = time.time()
+
+        for message in messages:
+            end = time.time()
+            if int(end - start) >= self.visibility_timeout:
+                # Don't add any more messages since they have
+                # re-appeared in the sqs queue Instead just reset and get
+                # fresh messages from the sqs queue
+                msg = (
+                    "Clearing Local messages since we exceeded "
+                    "their visibility_timeout"
+                )
+                logger.warning(msg)
+                break
```

Comment:
I know you asked about wanting both checks but I'm wondering - do we want either of them? My understanding about the initial motivation was that we wanted to avoid processing messages multiple times. The internal queue made it more likely that the message was pulled and the visibility timeout expired before we processed it. @spulec could speak to that more.

It's still possible to execute a message twice. Let's we have a batchsize of 2, a vis timeout of 60 seconds, and the first message takes 65 seconds. Message 2 could exceed the vis timeout without ever running it. 

However I don't know if the risk of running it multiple times is worse than the risk of pulling a message and never executing it. If we have a max receives of 1, that means message number 2 ends up in the DLQ without ever running it. IMO this is the largest pain point with pyqs today.

If we want to keep it at least now there's a way around it that is not available with the internal queue but it still sets up a potential trap.

Reply:
That is my understanding as well. It is there to avoid double processing a message. Would this not be relevant for example in sending email campaigns in clementine or are there application checks to avoid it. I am OK without doing this check and adding better documentation to explain the implications, and putting the onus on the user.

Reply:
I think most apps that have max receives > 1 should have logic to address if a message gets picked up again. I suppose it's possible the checks wouldn't capture if two are picked up at the exact same time. I think an at least once pattern is preferable to what we have here, which would lead to messages never being processed, but I don't know if that's true for all applications. 

The real issue here is batchsize. If we remove batchsize I don't think this check does anything (we read the message and immediately start processing it). Then can safely remove these checks and the possibility of something going straight to the DLQ without ever executing the task.